### ISSUE TITLE
Update front-end tooling

### ISSUE THREAD

**Issue by nmorduch** [wagtail/wagtail#5246]
Split from #5015, starting from https://github.com/wagtail/wagtail/issues/5015#issuecomment-482512717, where @thibaudcolas, @jonnyscholes, @lb-, and I have been discussing Prettier and stylelint rules.

**Comment by nmorduch** [comment]
In response to @lb-'s comment https://github.com/wagtail/wagtail/issues/5015#issuecomment-485566213

#### Re selector ordering

I don't love alphabetizing selectors, but I'm sufficiently convinced that an enforced ordering rule is beneficial for diffs. I personally tend to prefer orderings that group "rules that affect positioning/layout/sizing" and "rules that affect presentation", though I've never been very consistent on the details of that. 

I found [stylelint-config-idiomatic-order](https://github.com/ream88/stylelint-config-idiomatic-order) via the example config for prettier-stylelint, and I haven't tested it but it looks approximately like what I would like. You can see its specific ordering in its [index.js](https://github.com/ream88/stylelint-config-idiomatic-order/blob/master/index.js#L81). It looks like it outputs a config for stylelint-order that puts positioning, display, and box model first, then everything else is alphabetized after that. Any thoughts on how well that would work for us?

#### A thought on diffs for this change

This is probably already the plan, but I'd like to ask that, in the PR to add these rules, commits/diffs are set up in such a way that it's easy to distinguish "changes made by prettier-stylelint" and "changes made by humans". This probably mostly means putting any manual fixing of things that stylelint doesn't like into a separate commit from the enormous collection of automatic changes that prettier-stylelint will make the first time it is run.

**Comment by jonnyscholes** [comment]
Yeah I prefer idiomatic ordering over alphabetical too so would opt for that option if we enforce selector ordering. 

One grievance I do have with enforcing selector ordering though, is sometimes it's nice to be able to order your rules based on something specific to the design/functionality (eg themes or a complex CSS animation). For example, of late one thing I've been doing in parts of my code that have a lot of CSS transitions/animations is grouping all of the styles that are required only for the transition and are changed as part of it at the end of the rule. I've found this makes "designing" the animations much easier and ever so slightly closer to a visual timeline that one would use in something like After Effects (you can collapse all but the relevant code in your editor so you're only looking at "animation code").

Personally I wonder how much value enforcing ordering will bring? My grievance is pretty minor, but I guess I always try to make sure not to over saturate projects with aesthetic rules if possible and focus more on rules that stop people from breaking things.

**Comment by lb-** [comment]
@nmorduch 

**Selector Ordering**

That library looks good, I wish I had found that for our project, grouping some of the selectors makes a lot of sense.

It is still a matter of benefit vs hassle to maintain, I think the benefit is there and with the grouping will really aid in predictability/readability. If we can get our auto-formatting tooling to work with it well then I personally think it is worth it for the benefit for future developers and maintainers.

As per @jonnyscholes  - there are times where linting things like that can get in the way.

**Editor Guides**
Not sure if this is covered but it might be worth adding a sub-page about setting up your editor under http://docs.wagtail.io/en/v2.4/contributing/index.html

I personally use VS Code, used Atom before that.

**Enforcing existing rules in our docs**
Realised I should probably read what we have in our docs about this :) 

http://docs.wagtail.io/en/v2.4/contributing/css_guidelines.html

* sass imports - we can enforce this with  - `at-import-no-partial-leading-underscore` - https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/at-import-no-partial-leading-underscore/README.md
* pixels vs ems - can we look at enforcing rems for font size via `unit-blacklist` - https://stylelint.io/user-guide/rules/unit-blacklist/
* Specificity - I think rules covering those were already on @thibaudcolas  original comments
* Class naming conventions - docs will need to be changed as we are implementing a refined naming convention but we should be able to make a rule against using `js-` prefixed classnames via `selector-class-pattern`  https://stylelint.io/user-guide/rules/selector-class-pattern/

**Comment by lb-** [comment]
Proposed Next Action - Team meeting today, confirm agreement on creating a new repo for stylelint (and maybe prettier config) OR reworking the eslint repo into one with multiple folders and call it something like editor config. Prettier may not need it as it really does not need (or have) many config options.

This way we have a repo to start working with to draft config and test locally.

**Comment by nmorduch** [comment]
@jonnyscholes I imagine disabling linting the selector order for those styles, with a note as to why that was done, would work fine.

We should definitely update the CSS guidelines in general, and I like enforcing not using pixels except that I would make it so that things like borders can be in pixels (which is best for crisp lines)

**Comment by lb-** [comment]
As per yesterday's core team meeting, we agreed it would be good to isolate the stylelint config to its own repo (the same as eslint).

This has been created
https://github.com/wagtail/stylelint-config-wagtail

**Comment by lb-** [comment]
@nmorduch  + @thibaudcolas  - I had a chance to spend a bit of time on this today and have set up a base project with the following

* converted existing yaml config to index.js + confirmed that each rule is applicable in the latest version of Stylelint
* set up basic testing (I do not think just has to be extensive for every rule but just a sub-set plus it could help as a reference of good and bad scss in the future)
* set up eslint, prettier and stylelint for the project itself
* set up package.json, not published as I do not have access to NPM (but this should wait until we are ready)
* rough update to readme
* npm scripts

Happy for anything to change, I did my best to look at other Wagtail projects to see what conventions we follow for npm scripts, linting etc.

https://github.com/wagtail/stylelint-config-wagtail

Next Actions:

* Test out integration with Wagtail itself
* Add the additional rules / extends as per comments
* Maybe add a bit of a goal explanation to the readme, plus links to ITCSS and BEM for context
* Publish to NPM as a 0.0.1 release

**Comment by thibaudcolas** [comment]
That's great @lb-! I'm really happy to see this is using stylelint v10, even though it's only been out for a few days. I think we should look into using this config for Wagtail as soon as possible and not try to get the perfect config beforehand, so we can see the progress and more easily test addition of new rules.

Here's my general feedback, let me know if I can help by reviewing specific things.

- I don't think the config should use `eslint-config-wagtail` for its own linting, since this is outdated and tailored towards React / ES6 in the browser anyway. Just use the latest ESLint and its recommended config, with the correct env configuration.
- I think it would be quite cool for that project's README to be the one source of truth when it comes to "how we write CSS for Wagtail". This way the principles and the actual enforcement are all together in one place, and we can have discussions about CSS guidelines as issues on the config repo.
- Please give me your npm username so I can add you to https://www.npmjs.com/settings/wagtail/members (others welcome too)
---

There are further nice-to-haves I've set up / considered for my own config and would want to propose for future inclusion:

- Automated releases in CI with a tool like https://github.com/semantic-release/semantic-release. This makes it much easier to maintain "small" packages over time.
- Automated dependencies update with a tool like https://renovatebot.com/. Of a subset of the config's dependencies, again to make long-term maintenance easier.
- Automated self-documentation / export of the config rules in its README, so we can easily click on the links to understand what rules do (here's an [example](https://github.com/thibaudcolas/stylelint-config-cookbook#cookbook-config))
- Cross-browser support linting with https://github.com/ismay/stylelint-no-unsupported-browser-features
- Accessibility-related checks with https://github.com/YozhikM/stylelint-a11y

**Comment by lb-** [comment]
@thibaudcolas 

A few quick answers, I will work through this a bit more over the weekend if I get the chance.

- npm username is `lb`
- I personally would like to try to use `eslint-config-wagtail`, I am happy to maybe look at migrating the existing rules to latest (some will be deprecated) and then this could constitute a major version bump fro this package. Even though some rules will be jsx based, I think there is still merit in having a shared way of writing JS across all Wagtail projects.
- Love the idea of the Readme being used for that, only concern is that we will end up having duplicate content in the documentation also, there may be an automated way to merge these every now and then.
- all the other items look good, I will endeavour to get my head around them.
- I might add one suggestion, for smaller packages like this, it might be nice to use some semi-auto publishing process, I have seen other projects use [npmpub](https://github.com/MoOx/npmpub) - which could automate the parts of pre/post publish (eg. running tests).

**Comment by thibaudcolas** [comment]
👌 if you think it's a good time to update the ESLint config, then good idea!

For `npmpub`, yes, this sounds really similar to `semantic-release`. There are a few differences:

- semantic-release is meant to run in CI, to reduce the risk of human error as much as possible. 
- It's usually used to generate the CHANGELOG and determine the version number based on the commits, to further reduce the amount of manual work involved in a release.

Basically you only have to annotate commits in a predefined format, then releases happen without any further work than merging what's meant to be released. Perhaps not for every project / team, but I'd highly recommend it for shared configs.

**Comment by lb-** [comment]
@thibaudcolas  + @nmorduch + @jonnyscholes 

**Made a bunch of updates to our WIP stylelint-config project**

https://github.com/wagtail/stylelint-config-wagtail

- eslintrc env corrected
- added some additional documentation for contributing
- added nvmrc usage
- refined editor/prettier config (trying my best to match other Wagtail JS projects but they are all quite different)
- I have tested this locally and it appears to lint correctly and work well ([see code](https://github.com/wagtail/wagtail/compare/master...lb-:poc/stylelint-config-wagtail-local?expand=1)) YAY (no extra rules yet though)

**eslint usage**

- I have made some headway to getting the eslint project up to date in [this PR](https://github.com/wagtail/eslint-config-wagtail/pull/3)

**some items still to do**

- Have not added `semantic-release` as I am still getting my head around how that all works
- Renovatebot, same still getting my head around that
- Feel free to do these (or anyone else that wants to pitch in), I will aim to take another look this week

**Actual configuration**

- probably best we compile all the desired additions based on this thread and the previous and then add those in the normal PR format (I am just pushing to master currently).
- I think it would be nice to try to have our first NPM release (0.1.0) match the baseline (current) config, and then add on top of that with a next version bump (even 1.0.0).

**Comment by thibaudcolas** [comment]
It's great to see this happening @lb-! I think you should make a PR to Wagtail switching to the 0.1.0 of this config right away. The `semantic-release` and Renovate setups are very valuable, but not as much as having an updated config.

**Comment by lb-** [comment]
@thibaudcolas - cool. I will get on it. I have not seen a notification about the npm organisation. Could you please add me so I can publish the package.

**Comment by thibaudcolas** [comment]
My bad, done!

**Comment by thibaudcolas** [comment]
#5281 is in, thank you @lb- 🙂 

For the purpose of this specific issue, I think the next steps are to:

- Add more rules to make the config stricter
- Set up Prettier to format code automatically

For Prettier I think it's just a matter of someone making it happen, following from @lb-’s thoughts on how to do this in https://github.com/wagtail/wagtail/issues/5015#issuecomment-485566213. Would you be keen to take this forward @lb-?

For the additional rules, does anyone have thoughts on how to proceed?

**Comment by lb-** [comment]
@thibaudcolas  - happy to keep going with this process.

It would be really good to see if we can get the npm ownership of the eslint config project so we can also get that updated correctly.

But for now I will spend some time summarising the thoughts so far into a proposed set of rules and next steps.

**Comment by thibaudcolas** [comment]
I’m going to close this now that we’ve made significant progress on this, and it feels like separate issues might be more valuable.

- Further work on stylelint config: https://github.com/wagtail/stylelint-config-wagtail/issues/6
- ESLint config – merged @lb-’s https://github.com/wagtail/eslint-config-wagtail/pull/3 and https://github.com/wagtail/eslint-config-wagtail/pull/5, now I think the next step is https://github.com/wagtail/wagtail/issues/2797.
- Prettier #6059
